### PR TITLE
Numark Mixtrack Platinum midi mapping

### PR DESCRIPTION
**Pull request by iamcodemaker**
This is a mapping for the Numark Mixtrack Platinum controller. Nearly all of the features of the controller are supported including the LCD display screens and jog wheels. Support for controlling Mixxx's effects and effect parameters is supported as well. I have also extended the functionality of the controller to support 8 hotcues.

Documentation here https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum

There are a bunch of commits here, I can squash if necessary.

### CODE REVIEW

**Comment by radusuciu** [comment]
Looks like a very comprehensive mapping. No glaring issues popped up during my reading of this code. I would recommend looking at this though: https://www.codefactor.io/repository/github/mixxxdj/mixxx/pull/1855. You can safely ignore the "variable declared not used" warnings imo, but you may want to think about the code duplication issues.

Minor nitpick, some mixing of camelCase with under_score, but I don't think this should hold up merge.

Has this been tested by others?

**Comment by Be-ing** [comment]
Thanks for submitting this @iamcodemaker! And thanks for the review comments @radusuciu. I quickly glanced at the code and it looks neat and tidy. I'll go through the wiki and review the design for this soon, but right now my main focus is on fixing bugs for the 2.2.0 release. If we don't get it all done by the 2.2.0 release it's no big deal because we can easily make point releases.

**Comment by iamcodemaker** [comment]
I think I fixed the soft takeover issue on the EQ knobs. I'll need to test some more to be sure. Also there is another known issue: the first platter touch doesn't activate scratch mode properly for some reason. Things work fine on subsequent touches.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Not strictly necessary, but a good safety measure
```suggestion
                    if (component.tap !== undefined && typeof component.tap === 'function') {
```

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Is this necessary?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'm not sure this should toggle `show_parameters`. I would presume that mostly users with this controller would be using effect focusing to control the metaknob with the touchstrip, in which case showing all the parameters would be an annoying waste of space. On the other hand, I could imagine that a few users would be using the mixer knobs to adjust the individual parameters when an effect is focused. I'm a bit hesitant to add yet another option for this...

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Why not just use the default shift behavior of `components.CueButton`?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Clever way of implementing this :+1:

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: You can use the `script.toggleControl` shorthand.
```suggestion
            script.toggleControl(this.group, "keylock");
```

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I think this is more complicated than it needs to be. Why not just use `pitch_up/down`?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: There are two "hotcue button 8"s here. Did you accidentally do an extra copy & paste?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Why is this needed? Doesn't `loop_halve` with the default toggle Button behavior work? Likewise for `loop_double` below.

**Review by Be-ing — CHANGES_REQUESTED** [review]
I haven't read through all the code yet, but here's a start.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: It works, but it doesn't change the loop length for non-quantized loops. Setting `loop_scale` works for quantized and non-quantized loops. In fact the revision where `loop_halve` was changed to only work for quantized loops mentions this as a work around.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: An artifact of my personal coding style. For single line if statements I put everything on one line and omit the brackets. Being on one line makes it harder to accidentally add another statement with the same indent but leaving it out the if statement. I'll update it (and other instances) according to your style.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Serato. Will update.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Initially when I wrote this, I don't think the component system existed (or I wasn't aware of it). This code was copied over from that implementation. Some of these don't have components associated with them. It would be somewhat of a hassle to go through and figure out what's what and I don't see any benefit to it.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: probably not, I'll investigate

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I was on the fence about this. When I focus an effect, I use it to change the parameters. Though if you don't do that, I can see this being an issue. The only way around it would be adding an option. Side note: Maybe we need a UI for controller mapping options. That's something I'm willing to work on.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Always doing it one way means you don't have to worry about such subtle bugs and quirks of the language :)

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Okay, leave as it is then

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Okay, let's make it an option then. I agree, we should have a GUI for controller mapping options. But before making major changes to the mapping system I think we should finish #1795.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: When I look at the icon printed on the controller, I would expect the default `start_stop` behavior. I don't really understand the use case for not stopping playback with this. If you want to jump to the start and keep playing, you can use a normal hotcue.

> The reasoning here is that most people who get this controller and use it with mixxx won't be familar with mixxx.

I don't think that's a reliable assumption. Many Mixxx users do not have (much) experience with other DJ software, especially Linux users.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'm confused. That's what `sendShifted` is for and it's only hotcue 8 that has duplicate buttons here.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'm ok with going with the `start_stop` default. Thinking this through, jumping to the start and continuing to play doesn't seem very useful. Shift+play will jump to the cue point and I can't think of a reason you would want to jump to the start of the track instead of the cue point and keep playing. I always use this button to jump to the start when not playing the track, so making this the default would actually save me a button press.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Updated. I left setting `show_focus` enabled unconditionally, what are your thoughts on that? Should that also fall under the option to show the parameters?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Oh right, I forgot the context that this is in manual loop mode. Also, this mapping does not use `beatloop_toggle`, so updating `beatloop_size` doesn't really matter much. So I guess it is okay to leave it as is here.

Maybe we should introduce new ControlPushButtons for this use case instead of the unusual `loop_scale` Control. Figuring out how loop halve/double buttons should work when designing the current looping interface was very difficult and there may still be room for improvement. If you want to dig into the weeds of the discussion, it was in #1187.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Yes, `show_focus` should be toggled regardless of the option. There should not be a case where there is focus but the user can't see it.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Okay, please add a comment explaining this.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I believe there is one case, if you exit mixxx with an effect focused, that effect will be focused when you start again (the controller button will flash), but the ui will not show it as focused. I'd have to double check. This is something of a weird edge case though.

Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Hmm, if you can reproduce it try unplugging your controller before starting Mixxx again. There may be something in the mapping causing that.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I added a note to the [wiki page](https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum#manual_loop_mode) documenting this.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Oh, good catch. I think that is because the `_up` and `_down` ControlPushButtons are [automatically generated](https://github.com/mixxxdj/mixxx/blob/master/src/control/controlpotmeter.cpp#L60) from the [pitch ControlPotmeter](https://github.com/mixxxdj/mixxx/blob/master/src/engine/keycontrol.cpp#L24). In this case I think it would be helpful to specifically document them though. I only found them by looking at tooltips for the buttons in Deere in Developer Mode. Could you add them to the wiki?

Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: lol, was hoping you would do it. I'll take care of it.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: > Users could also just unplug the controller when they are done

Well, that's what I found so annoying with the Hercules P32, which has no way to disable the light show. I had to go through the extra step of unplugging it when I was just going to leave it sitting on my desk. That also made it quite annoying to quickly test something in Mixxx.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Thanks

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I figure most users won't even know about the extra options and in that case, the controller will work very similar to how the manual it comes with says it should. I tried to make all the defaults unsurprising. But when I play, I rarely use the autoloop functionality, so for me it makes sense to make that control cue points by default.

For the sampler, it probably makes sense to control 8 banks by default, but this might be unexpected to a user, so I want them to opt into it explicitly.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Unfortunately, there is no way to do this configuration. The controller sends the same midi codes regardless of the deck you are using to activate the sampler pads. We can tell when you switch into sampler mode on a deck, but the button presses aren't associated with the deck.

Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: :man_facepalming: :man_facepalming: why do manufacturers do these things...

Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Noted on the wiki https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum#sample_mode

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Should this array be initialized with 0s?

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I explained this above, but in short, doing this in `shift/unshift` can cause a "phantom" soft takeover mode. `ignoreNextValue()` gets called by `disconnect()` even if the control in question never moved every time shift is pressed. This strategy causes `ignoreNextValue()` to only get called if the control moved.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Okay, that's what I thought the issue might be. That might explain some issues I was having trouble reproducing when using my Hercules P32 mapping. So the implementation of the generic `Pot` in the library should be fixed.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Oh, I got confused because I misread the code and thought the `MixtrackPlatinum.encodeNum` function was being assigned instead of invoked.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I updated the [scripting wiki page](https://mixxx.org/wiki/doku.php/midi_scripting#soft-takeover) to document this. Thanks for pointing it out.

Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I noticed a related [bug in soft takeover](https://bugs.launchpad.net/mixxx/+bug/1805049).

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [inline res/controllers/Numark-Mixtrack-Platinum-scripts.js]: How does this handle showing time remaining? Or do you just need to send the controller firmware the time elapsed and duration, then it calculates remaining if you set it to show remaining?

**Review by Be-ing — CHANGES_REQUESTED** [review]

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: ping

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Yup. Tested this again. If you exit with an effect focused, it remains focused when you start again, but the ui doesn't show it. Should I make the mapping check for this on startup and enable `show_focus`?

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Haven't forgotten about this.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'll make this change, but I actually think all this indirection makes the math harder to read.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Right. We tell the controller the duration and we send the current time elapsed. The controller shows remaining/elapsed depending on which one it has been told to do.

Comment (Be-ing) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Okay, please add a comment explaining that.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'll make a sweep through and expand these soon.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: I'll make a sweep and fix these soon.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Going to leave this as is. I don't think it's a major concern for most users.

File: res/controllers/Numark-Mixtrack-Platinum-scripts.js
Comment (iamcodemaker) [reply res/controllers/Numark-Mixtrack-Platinum-scripts.js]: Going to leave as is. This will be less of an issue when we have a UI for controller options. The code will be just as complex, but the burden on the user will be less. I still think having all the options available is useful for users who want to customize their workflow.

**Comment by Be-ing** [comment]
I was about to merge this but I noticed it was targeted at master. Can you rebase this on 2.2 before merging?

I had a chance to play with a Mixtrack Platinum with Serato for a little bit a few weeks ago and I think it's a great deal and a good controller to recommend for new users. Thanks for your work on this!

**Comment by iamcodemaker** [comment]
Rebased. I was still planing on going through and tweaking the formatting of if/else statements and such, though it may be a few days before I can get to that. Obviously you can merge with out that.

**Pull request merged**
